cli/data.go

```diff
@@ -35,89 +37,182 @@ func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
-			return err
+		// If no data is returned, there is no more data.
+		if len(resp.GetData()) == 0 {
+			return nil
 		}
+		last = resp.GetLast()
 
-		gzippedBytes := datum.GetBinary()
-		r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
-		if err != nil {
-			return err
+		for _, bd := range resp.GetData() {
+			ids <- bd.GetMetadata().GetId()
 		}
+	}
+}
 
-		//nolint:gosec
-		dataFile, err := os.Create(filepath.Join(dst, dataDir, fileName+datum.GetMetadata().GetFileExt()))
-		if err != nil {
-			return errors.Wrapf(err, fmt.Sprintf("error creating file for file %s", datum.GetMetadata().GetId()))
-		}
-		//nolint:gosec
-		if _, err := io.Copy(dataFile, r); err != nil {
-			return err
-		}
-		if err := r.Close(); err != nil {
-			return err
-		}
-		numFilesDownloaded++
-		if numFilesDownloaded%logEveryN == 0 {
-			fmt.Fprintf(c.c.App.Writer, "downloaded %d files\n", numFilesDownloaded)
+func downloadBinary(ctx context.Context, client datapb.DataServiceClient, dst, id string) error {
+	var resp *datapb.BinaryDataByIDsResponse
+	var err error
+	for count := 0; count < maxRetryCount; count++ {
+		resp, err = client.BinaryDataByIDs(ctx, &datapb.BinaryDataByIDsRequest{
+			FileIds:       []string{id},
+			IncludeBinary: true,
+		})
+		if err == nil {
+			break
 		}
 	}
-	if numFilesDownloaded%logEveryN != 0 {
-		fmt.Fprintf(c.c.App.Writer, "downloaded %d files to %s\n", numFilesDownloaded, dst)
+	if err != nil {
+		return errors.Wrapf(err, "received error from server")
 	}
-	return nil
-}
+	data := resp.GetData()
 
-func (c *AppClient) sendBinaryDataByFilterRequest(filter *datapb.Filter, last string) (*datapb.BinaryDataByFilterResponse, error) {
-	return c.dataClient.BinaryDataByFilter(context.Background(), &datapb.BinaryDataByFilterRequest{
-		DataRequest: &datapb.DataRequest{
-			Filter: filter,
-			Limit:  1,
-			Last:   last,
-		},
-		IncludeBinary: true,
-		CountOnly:     false,
-	})
+	if len(data) != 1 {
+		return errors.Errorf("expected a single response, received %d", len(data))
+	}
+
+	datum := data[0]
+	mdJSONBytes, err := protojson.Marshal(datum.GetMetadata())
+	if err != nil {
+		return err
+	}
+
+	//nolint:gosec
+	jsonFile, err := os.Create(filepath.Join(dst, "metadata", datum.GetMetadata().GetId()+".json"))
+	if err != nil {
+		return err
+	}
+	if _, err := jsonFile.Write(mdJSONBytes); err != nil {
+		return err
+	}
+
+	gzippedBytes := datum.GetBinary()
+	r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
+	if err != nil {
+		return err
+	}
+
+	//nolint:gosec
+	dataFile, err := os.Create(filepath.Join(dst, "data", datum.GetMetadata().GetId()+datum.GetMetadata().GetFileExt()))
+	if err != nil {
+		return errors.Wrapf(err, fmt.Sprintf("error creating file for datum %s", datum.GetMetadata().GetId()))
+	}
+	//nolint:gosec
+	if _, err := io.Copy(dataFile, r); err != nil {
+		return err
+	}
+	if err := r.Close(); err != nil {
+		return err
+	}
+	return nil
 }
 
 // TabularData downloads binary data matching filter to dst.
```

Comment:
Looks like TabularData needs to sync to the newer changes with maxRetryCount, logging, and dataDir/metadataDir

Reply:
Oof yup more bad merging on my part, fixed